src/session.rs

```diff
@@ -659,54 +670,29 @@ where
                     debug!("proto [{}] has been closed", proto_id);
                 }
             }
-            _ => (),
-        }
-        self.distribute_to_substream(cx);
-    }
-
-    fn poll_inner_socket(&mut self, cx: &mut Context) -> Poll<Option<()>> {
-        if !self.state.is_normal() {
-            return Poll::Ready(None);
-        }
-        match Pin::new(&mut self.socket).as_mut().poll_next(cx) {
-            Poll::Ready(Some(Ok(sub_stream))) => {
-                self.handle_sub_stream(sub_stream);
-                Poll::Ready(Some(()))
-            }
-            Poll::Ready(None) => {
-                self.state = SessionState::RemoteClose;
-                Poll::Ready(None)
-            }
-            Poll::Pending => Poll::Pending,
-            Poll::Ready(Some(Err(err))) => {
-                debug!("session poll error: {:?}", err);
-                self.write_buf.clear();
-                self.high_write_buf.clear();
-                if !self.keep_buffer {
-                    self.read_buf.clear()
-                }
-
-                match err.kind() {
-                    ErrorKind::BrokenPipe
-                    | ErrorKind::ConnectionAborted
-                    | ErrorKind::ConnectionReset
-                    | ErrorKind::NotConnected
-                    | ErrorKind::UnexpectedEof => self.state = SessionState::RemoteClose,
-                    _ => {
-                        debug!("MuxerError: {:?}", err);
+            SessionEvent::StreamStart { stream } => self.handle_sub_stream(stream),
+            SessionEvent::ChangeState { state, error } => {
+                if self.state == SessionState::Normal {
+                    self.state = state;
```

Comment:
Should update state for all cases(not just `SessionState::Normal`)?

Reply:
no, if state is `LocalClose` or `Abnormal`, should not update